### ISSUE TITLE
Write operations over two pages is probably done incorrectly

### ISSUE THREAD

**Issue by jbaumg93** [panda-re/panda#441]
Hi! I observed a strange behavior with the taint2 plugin. When 4 bytes are written to the memory at once and the write operation is happing over two pages, then for the last bytes the consecutive physical addresses are labeled instead of the next page.

So for example, with 4kb pages, if a program writes 4 bytes to the virtual address 0x1ffe, then the operation affects all addresses up to 0x2002. The first two bytes are labeled correctly but the last two bytes end up on the next physical memory addresses. However, those are mapped by a different page, which means the wrong bytes are tainted.

To sum up, I think taint2 gets the physical address of the virtual one and labels all consecutive bytes in the physical memory space. However, this is not always correct, in case the write operation is performed over two pages because the latter bytes are translated to a different place in the physical address space.

I hope my explanation is somewhat understandable. Is this a known problem or do I get anything wrong?

**Comment by nathanjackson** [comment]
Do you have a recording that you can share that demonstrates the problem?

**Comment by jbaumg93** [comment]
Sure, you can find it [here](https://drive.google.com/file/d/1-eLWr_wmDQwpYJwY9CGLOssiNi8q2432/view?usp=sharing). I also hacked together a small test plugin, which should demonstrate the issue. It produces the following output (I annotated some lines):

```
Output buf of callback:
VMAWC buf phy: 0x0a3fbffe (vir: 0x04b83ffe): 0x30
VMAWC buf phy: 0x0a3fbfff (vir: 0x04b83fff): 0x78
VMAWC buf phy: 0x0bd17000 (vir: 0x04b84000): 0x30  <- Correct "jump" of physical address (another page)
VMAWC buf phy: 0x0bd17001 (vir: 0x04b84001): 0x2c
Output buf2, consecutive physical memory starting with 0xa3fbffe:
VMAWC buf2 phy: 0x0a3fbffe: 0x30
VMAWC buf2 phy: 0x0a3fbfff: 0x78
VMAWC buf2 phy: 0x0a3fc000: 0x08  <- Consecutive bytes contain 0x08 and 0x00
VMAWC buf2 phy: 0x0a3fc001: 0x00
taint_change phy: 0x0a3fbffe  size: 1
Mem: 0x30
Labels: 0x00000666
taint_change phy: 0x0a3fbfff  size: 1
Mem: 0x78
Labels: 0x00000666
taint_change phy: 0x0a3fc000  size: 1  <- Should be 0x0bd17000 and 0x30
Mem: 0x08
Labels: 0x00000666
taint_change phy: 0x0a3fc001  size: 1  <- Should be 0x0bd17001 and 0x2c
Mem: 0x00
Labels: 0x00000666
```

The output consists of three parts, the first one outputs buf of the callback after virtual memory write. You can see that the 4 bytes written uses two pages. The middle part shows the consecutive bytes 0x08 and 0x00. In the last part the taint change callback is used to show that the consecutive bytes 0x08 and 0x00 are tainted instead of bytes from another page (0x0bd17000 (+1): 0x30 and 0x2c).
